exercises/linked-list/src/lib.rs

```diff
@@ -0,0 +1,74 @@
+pub struct LinkedList<T>(std::marker::PhantomData<T>);
+
+pub struct Cursor<'a, T: 'a>(std::marker::PhantomData<&'a mut T>);
+
+pub struct Iter<'a, T: 'a>(std::marker::PhantomData<&'a T>);
+
+impl<T> LinkedList<T> {
+    pub fn new() -> Self {
+        unimplemented!()
+    }
+
+    pub fn len(&self) -> usize {
+        unimplemented!()
+    }
+
+    /// Return a cursor positioned at the tail node
+    pub fn cursor_tail(&mut self) -> Cursor<'_, T> {
+        unimplemented!()
+    }
+
+    /// Return a cursor positioned at the head node
+    pub fn cursor_head(&mut self) -> Cursor<'_, T> {
+        unimplemented!()
+    }
+
+    pub fn iter(&self) -> Iter<'_, T> {
+        unimplemented!()
+    }
+}
+
+impl<T> Drop for LinkedList<T> {
+    fn drop(&mut self) {
+        unimplemented!()
+    }
+}
+
+// the cursor is expected to act as if it is at the position of a node
+impl<'a, T: 'a> Cursor<'a, T> {
+    /// Take a mutable reference to the current element
+    pub fn peek_mut(&mut self) -> Option<&mut T> {
+        unimplemented!()
+    }
+
+    /// Move one position forward and return a reference to the new position
+    pub fn next(&mut self) -> Option<&mut T> {
+        unimplemented!()
+    }
+
+    /// Move one position backward and return a reference to the new position
+    pub fn prev(&mut self) -> Option<&mut T> {
+        unimplemented!()
+    }
+
+    /// Remove the element at the current node from the list and return it.
+    pub fn take(&mut self) -> Option<T> {
```

Comment:
What is the position of the cursor after `take`? That is, if I have the list `[A B C]`, and the cursor is at `B`, after `take`, where should the cursor be? This should be clearly specified in the comment.

Reply:
Policy options:

- cursor always moves forwards — what about when we `take` the last element in the list?
- cursor always moves backwards — what about when we `take` the first element in the list?
- cursor moves forwards, except when it is at the last element, in which case it moves backwards — what about when we `take` the final element in the list, leaving no elements remaining? 
- `take` consumes `self`, meaning we don't have to worry about the position of the cursor because the cursor dies when `take` is called

Of these options, I think I prefer the last.

If any of the first three options is chosen, there should be a test which exercises the complicated case.

Reply:
It's number 3. I was thinking in terms of indices like with a vec. The cursor stays in place and everything after is backshifted. If it's at the end, it moves to stay at the back. The cursor has an optional pointer precisely for the case of an empty list.  
There are valid use cases for `take and move left` or `take and move right` to minimize pointer chasing, but only one is required for generality. The other one can be gotten by stepping right after.  
My cursor here is conceptually different from the one in the `linked-list` crate. That one is located between elements and can be both before and after the list. This one is always conceptually on a node, which is why it moves backwards to follow the list when it's already at the end.

The last option precludes removal of several nodes. Traversing a list is costly, so that is not possible for a cursor.

Reply:
That makes sense, but if a cursor can shift in response to `take`, then that argues in favor of a single cursor which mutably shifts over the list instead of an immutable cursor per node. If we actually wanted the latter, then there's a fifth option which makes more sense than any of these: the cursor retains its pointer to the current element, which becomes the first entry in a new list.